src/gui/editors/PianoRoll.cpp

```diff
@@ -3884,16 +3894,14 @@ void PianoRoll::updateYScroll()
 						height() - PR_TOP_MARGIN -
 						SCROLLBAR_SIZE);
 
-	int total_pixels = m_octaveHeight * NumOctaves - (height() -
-					PR_TOP_MARGIN - PR_BOTTOM_MARGIN -
-							m_notesEditHeight);
-	m_totalKeysToScroll = qMax(0, total_pixels * KeysPerOctave / m_octaveHeight);
+	const int visible_space = height() - PR_TOP_MARGIN - PR_BOTTOM_MARGIN - m_notesEditHeight;
+	m_totalKeysToScroll = qMax(0, NumKeys - 1 - visible_space / m_keyLineHeight);
```

Comment:
Was the previous code not right?

Reply:
IIRC, the old code used something like `OctaveCount * KeysPerOctave`. That was incorrect because the MIDI specification has 128 keys, which is not divisible by 12, i.e. the last octave is incomplete. So the code needs to work wit `NumKeys` instead.

The other change (`visible_space` instead of `pixel_count`) was intended just to make the old code more readable / understandable.

Reply:
Ah! Gotcha, good catch. That might have been the source of the scrolling issue we'd been seeing.

Reply:
I would say the scrolling issues were not related to this: originally LMMS had only 96 notes, this PR extends it to support the full MIDI range. That's why the code needed an update; before that it was just fine.

Reply:
I meant the rounding error, that might have been why the last note would be hidden from scroll sometimes.